systemtest/src/test/java/io/strimzi/systemtest/cruisecontrol/CruiseControlST.java

```diff
@@ -203,6 +203,51 @@ void testCruiseControlWithRebalanceResourceAndRefreshAnnotation(ExtensionContext
+
+        resourceManager.createResourceWithoutWait(extensionContext, KafkaTemplates.kafkaWithCruiseControl(clusterName, 3, 1).build());
+
+        resourceManager.createResourceWithoutWait(extensionContext,  KafkaRebalanceTemplates.kafkaRebalance(clusterName)
+                .editMetadata()
+                .withNamespace(clusterOperator.getDeploymentNamespace())
+                .endMetadata()
+                .build());
+
+        KafkaRebalanceUtils.waitForKafkaRebalanceCustomResourceState(clusterOperator.getDeploymentNamespace(), clusterName, KafkaRebalanceState.ProposalReady);
+
+        KafkaRebalanceResource.replaceKafkaRebalanceResourceInSpecificNamespace(clusterName, kafkaRebalance -> kafkaRebalance.getSpec().setGoals(List.of("custom-goal")), clusterOperator.getDeploymentNamespace());
+
+        KafkaRebalanceUtils.waitForKafkaRebalanceCustomResourceState(clusterOperator.getDeploymentNamespace(), clusterName, KafkaRebalanceState.NotReady);
+
+        KafkaRebalanceStatus kafkaRebalanceStatus = KafkaRebalanceResource.kafkaRebalanceClient().inNamespace(clusterOperator.getDeploymentNamespace()).withName(clusterName).get().getStatus();
+        assertThat(kafkaRebalanceStatus.getConditions().get(0).getReason(), is("CruiseControlRestException"));
+    }
+
+    @IsolatedTest
+    void testCruiseControlChangesFromRebalancingtoProposalReadyWhenSpecUpdated(ExtensionContext extensionContext) {
+        String clusterName = mapWithClusterNames.get(extensionContext.getDisplayName());
+
+        resourceManager.createResourceWithoutWait(extensionContext, KafkaTemplates.kafkaWithCruiseControl(clusterName, 3, 1).build());
+
+        resourceManager.createResourceWithoutWait(extensionContext,  KafkaRebalanceTemplates.kafkaRebalance(clusterName)
+                .editMetadata()
+                .withNamespace(clusterOperator.getDeploymentNamespace())
+                .endMetadata()
+                .build());
+
+        KafkaRebalanceUtils.waitForKafkaRebalanceCustomResourceState(clusterOperator.getDeploymentNamespace(), clusterName, KafkaRebalanceState.ProposalReady);
+
+        LOGGER.info("Annotating KafkaRebalance: {} with 'approve' anno", clusterName);
+        KafkaRebalanceUtils.annotateKafkaRebalanceResource(new Reconciliation("test", KafkaRebalance.RESOURCE_KIND, clusterOperator.getDeploymentNamespace(), clusterName), clusterOperator.getDeploymentNamespace(), clusterName, KafkaRebalanceAnnotation.approve);
+        KafkaRebalanceUtils.waitForKafkaRebalanceCustomResourceState(clusterOperator.getDeploymentNamespace(), clusterName, KafkaRebalanceState.Rebalancing);
+
+        KafkaRebalanceResource.replaceKafkaRebalanceResourceInSpecificNamespace(clusterName, kafkaRebalance -> kafkaRebalance.getSpec().setReplicationThrottle(100000), clusterOperator.getDeploymentNamespace());
+
+        KafkaRebalanceUtils.waitForKafkaRebalanceCustomResourceState(clusterOperator.getDeploymentNamespace(), clusterName, KafkaRebalanceState.ProposalReady);
```

Comment:
Will this be really stable? It sounds like this will be open to race conditions where you miss the `Rebalancing` state.

Reply:
We could most definitely miss the `Rebalancing` state, maybe we should skip the wait on line 244 where we wait for the `Reblancing` state? I am not sure we can guarantee or force the resource will be in `Rebalancing` state when we check, or test this specific scenario reliably. That being said, if we remove the wait on line 244 the resource should end in the `ProposalReady` regardless the outcome of the race condition.

Reply:
Great idea @kyguy. Let me fix that